File: solutions/julia/nucleotide-count/1/nucleotide-count.jl
"""
    count_nucleotides(strand)

The count of each nucleotide within `strand` as a dictionary.

Invalid strands raise a `DomainError`.

"""

function count_nucleotides(strand::String) :: Dict{Char,Integer}
    result = Dict('A'=>0, 'C'=>0, 'G'=>0, 'T'=>0)
    for chr in strand
        if haskey(result, chr)
            setindex!(result, 1+(getindex(result, chr)), chr)
        else
            throw(DomainError(chr, "Invalid letter"))
        end
    end
    result
end

File: solutions/ruby/nucleotide-count/1/nucleotide_count.rb
class Nucleotide
  def self.from_dna(str)
    raise ArgumentError if str =~ /[^GTCA]/
    self.new(str)
  end
  
  def initialize(str)
    @str = str
  end

  def count(letter)
    @str.count(letter)
  end

  def histogram
    Hash[%w(A T C G).map{|ltr| [ltr, count(ltr)]}]
  end
end